sympy/solvers/solveset.py

```diff
@@ -1121,3 +1124,597 @@ def linsolve(system, *symbols):
+    because `linsolve` is better for all kind of linear system.
+
+    >>> nonlinsolve([x + 2*y -z - 3, x - y - 4*z + 9 , y + z - 4], [x, y, z])
+    {(3*z - 5, -z + 4, z)}
+
+    * system having polynomial equations:
+
+    >>> e1 = sqrt(x**2 + y**2) - 10
+    >>> e2 = sqrt(y**2 + (-x + 10)**2) - 3
+    >>> nonlinsolve((e1, e2), (x, y))
+    {(191/20, -3*sqrt(391)/20), (191/20, 3*sqrt(391)/20)}
+    >>> nonlinsolve([x**2 + 2/y - 2, x + y - 3], [x, y])
+    {(1, 2), (1 + sqrt(5), -sqrt(5) + 2), (-sqrt(5) + 1, 2 + sqrt(5))}
+    >>> nonlinsolve([x**2 + 2/y - 2, x + y - 3], [y, x])
+    {(2, 1), (2 + sqrt(5), -sqrt(5) + 1), (-sqrt(5) + 2, 1 + sqrt(5))}
+
+    How nonlinsolve is better than old solver `_solve_system` :
+    ===========================================================
+
+    * A positive dimensional system solver : nonlinsolve can return
+    solution for positive dimensional system. It finds the
+    Groebner Basis of the positive dimensional system(calling it as
+    basis) then we can start solving equation(having least number of
+    variable first in the basis) using solveset and substituting that
+    solved solutions into other equation(of basis) to get solution in
+    terms of minimum variables. Here the important thing is how we
+    are substituting the known values and in which equations.
+
+    * Real and Complex both solutions : nonlinsolve returns both real
+    and complex solution. If all the equations in the system are polynomial
+    then using `solve_poly_system` both real and complex solution is returned.
+    If all the equations in the system are not polynomial equation then goes to
+    `substitution` method with this polynomial and non polynomial equation(s),
+    to solve for unsolved variables. Here to solve for particular variable
+    solveset_real and solveset_complex is used. For both real and complex
+    solution function `_solve_using_know_values` is used inside `substitution`
+    function.(`substitution` function will be called when there is any non
+    polynomial equation(s) is present). When solution is valid then add its
+    general solution in the final result.
+
+    * Complements and Intersection will be added if any : nonlinsolve maintains
+    dict for complements and Intersections. If solveset find complements or/and
+    Intersection with any Interval or set during the execution of
+    `substitution` function ,then complement or/and Intersection for that
+    variable is added before returning final solution.
+
+    """
+    from sympy.polys.polytools import is_zero_dimensional, groebner
+
+    if not system:
+        return S.EmptySet
+
+    if not symbols:
+        raise ValueError('Symbols must be given, for which solution of the '
+                         'system is to be found.')
+
+    if hasattr(symbols[0], '__iter__'):
+        symbols = symbols[0]
+
+    try:
+        sym = symbols[0].is_Symbol
+    except AttributeError:
+        sym = False
+
```

Comment:
The example is in `nonlinsolve` docs (below). If only `[a, pi]` is passed as `symbols` then it will take other symbols(b, c, d) as const (user is not interested for these symbols) and value of ` pi` is `pi` .
Should we solve for all the variables present in the equations ?